internal/build/docker.go

```diff
@@ -189,79 +191,87 @@ func (l *localDockerBuilder) buildFromDfWithFiles(ctx context.Context, df Docker
 	return getDigestFromAux(*result)
 }
 
-func (l *localDockerBuilder) execStepsOnImage(ctx context.Context, img digest.Digest, steps []model.Cmd) (digest.Digest, error) {
-	imageWithSteps := img
-	for _, s := range steps {
-		cId, err := l.startContainer(ctx, containerConfigRunCmd(imageWithSteps, s))
-		if err != nil {
-			return "", fmt.Errorf("startContainer '%s': %v", img, err)
-		}
-
-		id, err := l.dcli.ContainerCommit(ctx, cId, types.ContainerCommitOptions{})
-		if err != nil {
-			return "", fmt.Errorf("containerCommit: %v", err)
-		}
-		imageWithSteps = digest.Digest(id.ID)
+func (l *localDockerBuilder) execInContainer(ctx context.Context, cID containerID, cmd model.Cmd) error {
+	created, err := l.dcli.ContainerExecCreate(ctx, string(cID), types.ExecConfig{
+		Cmd: cmd.Argv,
+	})
+	if err != nil {
+		return err
 	}
 
-	return digest.Digest(imageWithSteps), nil
-}
-
-// TODO(maia): can probs do this in a more efficient place -- e.g. `execStepsOnImage`
-// already spins up + commits a container, maybe piggyback off that?
-func (l *localDockerBuilder) imageWithEntrypoint(ctx context.Context, img digest.Digest, entrypoint model.Cmd) (digest.Digest, error) {
-	cId, err := l.startContainer(ctx, containerConfigRunCmd(img, model.Cmd{}))
-	if err != nil {
-		return "", fmt.Errorf("startContainer '%s': %v", string(img), err)
+	execID := execID(created.ID)
+	if execID == "" {
+		return fmt.Errorf("execInContainer: failed to create")
 	}
 
-	// Only attach the entrypoint if there's something in it
-	opts := types.ContainerCommitOptions{}
-	if !entrypoint.Empty() {
-		opts.Changes = []string{entrypoint.EntrypointStr()}
+	attached, err := l.dcli.ContainerExecAttach(ctx, string(execID), types.ExecStartCheck{})
+	if err != nil {
+		return fmt.Errorf("execInContainer#attach: %v", err)
 	}
+	defer attached.Close()
 
-	id, err := l.dcli.ContainerCommit(ctx, cId, opts)
+	// TODO(nick): feed this reader into the logger
+	buf := bytes.NewBuffer(nil)
+	_, err = io.Copy(buf, attached.Reader)
 	if err != nil {
-		return "", fmt.Errorf("containerCommit: %v", err)
+		return fmt.Errorf("execInContainer#copy: %v", err)
 	}
 
-	return digest.Digest(id.ID), nil
+	for true {
+		inspected, err := l.dcli.ContainerExecInspect(ctx, string(execID))
+		if err != nil {
+			return fmt.Errorf("execInContainer#inspect: %v", err)
+		}
+
+		if inspected.Running {
+			continue
+		}
+
+		status := inspected.ExitCode
+		if status != 0 {
+			return fmt.Errorf("Failed with exit code %d. Output:\n%s", status, buf.String())
```

Comment:
todo: option to print output even if exit code = 0 (verbose/debug logging?)

Reply:
there's already a TODO for that? before the loop

Reply:
ah whoops